src/sqlfluff/utils/reflow/depthmap.py

```diff
@@ -0,0 +1,128 @@
+"""The DepthMap class which is an enriched sequence of raw segments."""
+
+import logging
+from dataclasses import dataclass
+from typing import List, Sequence, Set, Tuple
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.parser.segments.raw import RawSegment
+
+
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class DepthInfo:
+    """An object to hold the depth information for a specific raw segment."""
+
+    stack_depth: int
+    stack_hashes: List[int]
+    # This is a convenience cache to speed up operations.
+    stack_hash_set: Set[int]
+    stack_class_types: List[Set[str]]
+
+    @classmethod
+    def from_raw_and_stack(cls, raw: RawSegment, stack: Sequence[BaseSegment]):
+        """Construct from a raw and it's stack."""
+        stack_hashes = [hash(seg) for seg in stack]
+        return cls(
+            stack_depth=len(stack),
+            stack_hashes=stack_hashes,
+            stack_hash_set=set(stack_hashes),
+            stack_class_types=[seg.class_types for seg in stack],
+        )
+
+    def common_with(self, other: "DepthInfo") -> List[int]:
+        """Get the common depth and hashes with the other."""
+        # We use set intersection because it's faster and hashes should be unique.
+        common_hashes = self.stack_hash_set.intersection(other.stack_hashes)
+        common_depth = len(common_hashes)
+        return self.stack_hashes[:common_depth]
+
+    def trim(self, amount: int):
+        """Return a DepthInfo object with some amount trimmed."""
+        if amount == 0:
+            # The trivial case.
+            return self
+        return self.__class__(
+            stack_depth=self.stack_depth - amount,
+            stack_hashes=self.stack_hashes[:-amount],
+            stack_hash_set=self.stack_hash_set.difference(self.stack_hashes[-amount:]),
+            stack_class_types=self.stack_class_types[:-amount],
+        )
+
+
+class DepthMap:
+    """A sequence raw segments with depth and parent information.
+
+    This class addresses two needs:
+    - To understand configuration of segments with no whitespace
+      within them - so the config is related to the parent and
+      not the segment)
+    - To map the depth of an indent points to apply some precedence
+      for where to insert line breaks.
+
+    The internals are structured around a list to do lookups
+    and a dict (keyed with the raw segment UUID) to hold the rest.
+
+    """
+
+    def __init__(self, raws_with_stack: Sequence[Tuple[RawSegment, List[BaseSegment]]]):
+        # TODO: decide whether we need the raw segments?
+        # self.raw_segments = []
+        self.depth_info = {}
+        for raw, stack in raws_with_stack:
+            # self.raw_segments.append(raw)
+            self.depth_info[raw.uuid] = DepthInfo.from_raw_and_stack(raw, stack)
+
+    @classmethod
+    def from_parent(cls, parent: BaseSegment):
```

Comment:
Does this work? (Found this pattern documented here: https://stackoverflow.com/questions/44640479/type-annotation-for-classmethod-returning-instance)
```suggestion
    def from_parent(cls: Type[T], parent: BaseSegment) -> T:
```

Reply:
it doesn't, but this does:
```python
def from_parent(cls: Type["DepthMap"], parent: BaseSegment) -> "DepthMap":
   ....
```